Author: Nicholas Bamber <nichoolas@periapt.co.uk>
Subject: convert to argp and add extra arguments and resulting bug fixes
New args
 --ident identity of the log helper process in syslog
 --pid pid file of the duende daemon
 --uid uid of the log helper process, defaults to 66
 --chroot directory of the log helper chroot, defaults to /etc/maradns/logger
 --gid gid of the log helper process, defaults to 66
 --restart_on_exit exit status, defaults to 8
Changed functionality
 Log helper chroots 
 Consolidated waitpid calls to avoid race condition and ensure child reaping
Last-Update: 2011-06-30
Forwarded: yes
--- a/tools/duende.c
+++ b/tools/duende.c
@@ -40,11 +40,78 @@
 #include <sys/types.h>
 #include <sys/wait.h>
 #include <unistd.h>
+#include <argp.h>
+#include <string.h>
 #include "../MaraDns.h"
 
 int got_hup_signal = 0;
 int got_term_signal = 0;
 
+/* needed for command line overrides */
+struct argument
+{
+    int duende_uid; /* = DUENDE_LOGGER_UID; */
+    int duende_gid; /* = DUENDE_LOGGER_UID; */
+    int restart_on_exit;
+    const char *duende_chroot;
+    const char *duende_pid_file;
+    const char *duende_ident;
+    int argc;
+    char **argv;
+};
+
+static const struct argp_option options[] = {
+    {"pid",    'p', "FILE", 0, "File used to store pid of duende process"},
+    {"uid",    'u', "UID",  0, "Userid log helper should change down to"},
+    {"gid",    'g', "GID",  0, "Groupid log helper should change down to"},
+    {"chroot", 'c', "DIR",  0, "Directory log helper should change and chroot to"},
+    {"ident",  'i', "STR",  0, "How log helper should be identified in syslog"},
+    {"restart_on",    'r', "INT",  0, "Exit status on which to restart child process"},
+    {0}
+};
+
+static error_t
+parse_opt(int key, char *arg, struct argp_state *state)
+{
+    struct argument *arguments = state->input;
+
+    switch(key) {
+        case 'c': /* chroot */
+            arguments->duende_chroot = arg;
+            break;
+        case 'i': /* ident */
+            arguments->duende_ident = arg;
+            break;
+        case 'u': /* uid */
+            arguments->duende_uid = atoi(arg);
+            break;
+        case 'r': /* restart on */
+            arguments->restart_on_exit = atoi(arg);
+            break;
+        case 'g': /* gid */
+            arguments->duende_gid = atoi(arg);
+            break;
+        case 'p': /* pid */
+            arguments->duende_pid_file = arg;
+            break;
+        case ARGP_KEY_NO_ARGS:
+            argp_usage(state);
+        case ARGP_KEY_ARGS:
+            arguments->argc = state->argc - state->next;
+            arguments->argv = state->argv + state->next;
+            break;
+        default:
+            return ARGP_ERR_UNKNOWN;
+    }
+
+    return 0;
+}
+
+static char doc[] = "duende -- a program to manage a daemon process, logging its standard output and error to syslog";
+static char args_doc[] = "command [arguments...]";
+
+static struct argp argp = {options, parse_opt, args_doc, doc};
+
 /* If we get a HUP signal set the flag so we can restart the MaraDNS child
    process */
 void handle_hup() {
@@ -59,7 +126,7 @@
 
 /* Helper process which syslogs stuff from either MaraDNS' stdout or stderr. */
 
-void log_helper(char *name,int stdout_fd) {
+void log_helper(const struct argument *arguments, int stdout_fd) {
     char out_buf[1024];
 
     /* We can't use our signal handlers because fgets is blocking */
@@ -67,22 +134,35 @@
     signal(SIGHUP,SIG_DFL);
 
     /* Open up the sys log */
-    openlog(name,0,LOG_DAEMON);
+    syslog(LOG_ALERT,"%s process starting up", arguments->duende_ident);
 
     /* Drop all privileges */
-    if(chdir(DUENDE_CHROOT_DIR) != 0) {
-       syslog(LOG_ALERT,"Can not enter chroot directory %s",DUENDE_CHROOT_DIR);
+    if (chdir(arguments->duende_chroot) != 0) {
+        syslog(LOG_ALERT,"Can not enter chroot directory %s", arguments->duende_chroot);
         syslog(LOG_ALERT,"%s","We can not log daemon output");
         printf("Fatal error logging; read syslog\n");
-        printf("%s directory required to exist\n",DUENDE_CHROOT_DIR);
+        printf("%s directory required to exist\n",arguments->duende_chroot);
         exit(1);
-        }
-    if(setuid(DUENDE_LOGGER_UID) != 0) {
-        syslog(LOG_ALERT,"%s%d","Can not change UID to ",DUENDE_LOGGER_UID);
+    }
+#if ! (defined __CYGWIN__ || defined QNX)
+    if (chroot(arguments->duende_chroot) != 0) {
+        syslog(LOG_ALERT,"Can not chroot to directory %s",arguments->duende_chroot);
+        printf("Fatal error logging; read syslog\n");
+        exit(1);
+    }
+#endif
+    if(setgid(arguments->duende_gid) != 0) {
+        syslog(LOG_ALERT,"%s%d","Can not change GID to ", arguments->duende_gid);
+        printf("Fatal error logging; read syslog\n");
+        exit(1);
+    }
+    if(setuid(arguments->duende_uid) != 0) {
+        syslog(LOG_ALERT,"%s%d","Can not change UID to ", arguments->duende_uid);
         syslog(LOG_ALERT,"%s","We can not log daemon output");
         printf("Fatal error logging; read syslog\n");
         exit(1);
-        }
+    }
+
 
     /* Log both stdout and stderr */
     dup2(stdout_fd,0);
@@ -94,7 +174,7 @@
              * an issue that can cause confusion */
             syslog(LOG_ALERT,"%s",out_buf);
 #else /* __FreeBSD__ */
-            syslog(LOG_INFO,"%s",out_buf);
+            syslog(LOG_ALERT,"%s",out_buf);
 #endif /* __FreeBSD__ */
         }
 
@@ -108,9 +188,9 @@
    a HUP signal.  In either of these cases, restart the child daemon and the
    (if needed) logger process).  Otherwise, exit */
 
-void handle_child_exited(int exit_status, pid_t alive, pid_t exited) {
+void handle_child_exited(int exit_status, pid_t alive, pid_t exited, int restart_code) {
         if(WIFEXITED(exit_status)) { /* Exit with exit code */
-            if(WEXITSTATUS(exit_status) != 8) { /* Anything but HUP */
+            if(WEXITSTATUS(exit_status) != restart_code) { /* Anything but HUP */
                 kill(alive,SIGTERM);
                 syslog(LOG_ALERT,"Child exited with status %d",exit_status);
                 exit(WEXITSTATUS(exit_status));
@@ -130,9 +210,9 @@
             kill(alive,SIGTERM);
             exit(2);
             }
-        /* Clean up the system logging */
-        syslog(LOG_ALERT,"Cleaning up system logging");
-        kill(alive,SIGTERM);
+       /* Clean up the system logging */
+       syslog(LOG_ALERT,"Cleaning up system logging");
+       kill(alive,SIGTERM);
     }
 
 /* The main process forks off the child.  Right now, I will just have
@@ -144,27 +224,33 @@
 
 int main(int argc, char **argv) {
     int exit_status;
+    int check_pid;
     pid_t pid, log_pid;
     int stream1[2]; /* Used for piping */
-    int exec_argv_offset = 1; /* Also used to determine PID writing */
-    if(argv[0] == NULL || argv[1] == NULL) {
-        printf("Usage: duende (--pid=/path/to/file) [program] [arguments]\n");
+    struct argument arguments;
+
+    /* Initialize data structures */
+    arguments.duende_chroot = DUENDE_CHROOT_DIR;
+    arguments.duende_pid_file = 0;
+    arguments.duende_gid = DUENDE_LOGGER_UID;
+    arguments.duende_uid = DUENDE_LOGGER_UID;
+    arguments.duende_ident = "log-helper";
+    arguments.restart_on_exit = 8;
+    arguments.argc = 0;
+    arguments.argv = 0;
+
+    argp_parse(&argp, argc, argv, ARGP_IN_ORDER, 0, &arguments);
+
+    if (arguments.argv == 0 || arguments.argv[0] == 0) {
+        printf("FATAL: Unable to identify command\n");
         exit(1);
-        }
-    if(!strncasecmp(argv[1],"--pid=",6)) {
-        if(argv[2] == NULL) {
-            printf(
-		"Usage: duende (--pid=/path/to/file) [program] [arguments]\n");
-            exit(1);
-            }
-        exec_argv_offset = 2;
-        }
+    }
 
     /* Let children know that duende is running */
     if(setenv("DUENDE_IS_RUNNING","1",0) != 0) {
         printf("FATAL: Unable to set environment variable\n");
         exit(1);
-        }
+    }
 
     /* The parent immediately exits */
     if(fork() != 0)
@@ -174,17 +260,16 @@
     setpgid(0,0); /* No longer visible in 'ps' without the 'auxw' argument */
 
     /* Write our PID to a file if the user so desires us to */
-    if(exec_argv_offset == 2) {
-        FILE *fp_pid = fopen(argv[1] + 6,"w");
-        if(!fp_pid) {
+    if(arguments.duende_pid_file) {
+        FILE *fp_pid = fopen(arguments.duende_pid_file,"w");
+        if (!fp_pid) {
             syslog(LOG_ALERT,"Fatal writing, to PID file, error\n");
             exit(1);
-            }
+        }
         unsigned int local_pid = getpid();
         fprintf(fp_pid,"%u",local_pid);
         fclose(fp_pid);
-        }
-
+    }
     /* Sysadmins expect HUP to reload, so we set that up */
     signal(SIGHUP,handle_hup);
     signal(SIGTERM,handle_term);
@@ -192,73 +277,77 @@
 
     pid = 0; log_pid = 0;
 
+    openlog(arguments.duende_ident,LOG_PID,LOG_DAEMON);
+
     for(;;) {
-        if(pipe(stream1) != 0) {
+        if (pipe(stream1) != 0) {
             syslog(LOG_ALERT,"Fatal pipe error");
             exit(3);
-            }
+        }
         pid = fork();
-        if(pid == -1) {
+        if (pid == -1) {
             syslog(LOG_ALERT,"Fatal pid error");
             exit(1);
-            }
-        if(pid == 0) { /* Child; this one execs maradns */
+        }
+        if (pid == 0) { /* Child; this one execs maradns */
             close(stream1[0]);
             /* Dup the standard output */
-            if(dup2(stream1[1],1) != 1) {
+            if (dup2(stream1[1],1) != 1) {
                 syslog(LOG_ALERT,"Fatal dup2 error 1");
                 exit(4);
-                }
+            }
             /* And the standard error */
-            if(dup2(stream1[1],2) != 2) {
+            if (dup2(stream1[1],2) != 2) {
                 syslog(LOG_ALERT,"Fatal dup2 error 2");
                 exit(5);
-                }
-            argv[0] = argv[exec_argv_offset];
-            execvp(argv[exec_argv_offset],argv + exec_argv_offset);
+            }
+            execvp(arguments.argv[0],arguments.argv);
             /* OK, not found */
-            printf("duende: %s: Command can't run, terminating\n",argv[exec_argv_offset]);
+            printf("duende: %s: Command can't run, terminating\n",arguments.argv[0]);
             syslog(LOG_ALERT,"Command can't run, terminating\n");
             exit(1);
-            }
+        }
 
         /* Parent */
         close(stream1[1]);
         log_pid = fork();
-        if(log_pid == 0) { /* Child to syslog all of MaraDNS' output */
-            argv[0] = "duende-log-helper";
-            log_helper(argv[exec_argv_offset],stream1[0]);
+        if (log_pid == 0) { /* Child to syslog all of MaraDNS' output */
+            log_helper(&arguments, stream1[0]);
             syslog(LOG_ALERT,"log_helper finished, terminating\n");
             exit(1);
-            }
+        }
         for(;;) {
             /* If we got a HUP signal, send it to the child */
-            if(got_hup_signal == 1) {
+            if (got_hup_signal == 1) {
                 kill(pid,SIGHUP);
                 got_hup_signal = 0;
-                }
+            }
             /* If we got a TERM or INT signal, send it to the children
                then exit ourselves */
-            else if(got_term_signal == 1) {
+            else if (got_term_signal == 1) {
                 /* XXX: make sure term really stops the children */
                 kill(pid,SIGTERM);
                 kill(log_pid,SIGTERM);
                 syslog(LOG_ALERT,"got term signal, terminating\n");
                 exit(0);
-                }
+            }
             sleep(1);
-            if(waitpid(pid,&exit_status,WNOHANG) == pid) { /* If child ended */
-                handle_child_exited(exit_status,log_pid,pid);
+            int wait_pid = waitpid(-1, &exit_status, WNOHANG);
+            if (wait_pid == pid) { /* If child ended */
+                handle_child_exited(exit_status,log_pid,pid,arguments.restart_on_exit);
                 close(stream1[0]);
                 break; /* Out of the inner loop; re-start Mara */
-                }
+            }
             /* If logger terminated */
-            if(waitpid(log_pid,&exit_status,WNOHANG) == log_pid) {
-                handle_child_exited(exit_status,pid,log_pid);
+            else if (wait_pid == log_pid) {
+                handle_child_exited(exit_status,pid,log_pid,arguments.restart_on_exit);
                 close(stream1[0]);
                 break; /* Out of the inner loop; re-start Mara */
-                }
+            }
+            else if (wait_pid > 0) {
+                syslog(LOG_ALERT, "unexpected child reaped: %i", wait_pid);
             }
         }
     }
+}
 
